cpp/dolfinx/fem/DirichletBC.cpp

```diff
@@ -427,56 +537,114 @@ fem::locate_dofs_topological(const fem::FunctionSpace& V, int dim,
+
+    if (remote)
     {
-      const int index = entity_dofs[entity_local_index][j];
-      dofs.push_back(cell_dofs[index]);
+      const std::vector dofs_remote
+          = get_remote_bcs1(*V.dofmap()->index_map, dofs);
+
+      // Add received bc indices to dofs_local
+      dofs.insert(dofs.end(), dofs_remote.begin(), dofs_remote.end());
+
+      // Remove duplicates
+      std::sort(dofs.begin(), dofs.end());
+      dofs.erase(std::unique(dofs.begin(), dofs.end()), dofs.end());
     }
   }
+  else
+  {
+    // V is a sub space we need to take the block size of the dofmap and index
+    // map into account, as they differ
+    const int bs = dofmap->bs();
+    const int element_bs = dofmap->element_dof_layout->block_size();
 
-  // TODO: is removing duplicates at this point worth the effort?
-  // Remove duplicates
-  std::sort(dofs.begin(), dofs.end());
-  dofs.erase(std::unique(dofs.begin(), dofs.end()), dofs.end());
+    // Iterate over marked facets
+    for (std::int32_t e : entities)
+    {
+      // Get first attached cell
+      assert(e_to_c->num_links(e) > 0);
+      const int cell = e_to_c->links(e)[0];
 
-  if (remote)
-  {
-    const std::vector dofs_remote
-        = get_remote_bcs1(*V.dofmap()->index_map, dofs);
+      // Get local index of facet with respect to the cell
+      auto entities_d = c_to_e->links(cell);
+      auto it = std::find(entities_d.begin(), entities_d.end(), e);
+      assert(it != entities_d.end());
+      const int entity_local_index = std::distance(entities_d.begin(), it);
 
-    // Add received bc indices to dofs_local
-    dofs.insert(dofs.end(), dofs_remote.begin(), dofs_remote.end());
+      // Get cell dofmap
+      xtl::span<const std::int32_t> cell_dofs = dofmap->cell_dofs(cell);
 
+      // Loop over facet dofs and 'unpack' blocked dofs
+      for (int i = 0; i < num_entity_closure_dofs; ++i)
+      {
+        const int index = entity_dofs[entity_local_index][i];
+        for (int block = 0; block < element_bs; ++block)
+        {
+          const std::div_t pos = std::div(element_bs * index + block, bs);
+          dofs.push_back(bs * cell_dofs[pos.quot] + pos.rem);
+        }
+      }
+    }
+
+    // TODO: is removing duplicates at this point worth the effort?
     // Remove duplicates
     std::sort(dofs.begin(), dofs.end());
     dofs.erase(std::unique(dofs.begin(), dofs.end()), dofs.end());
-  }
 
+    if (remote)
+    {
+      // Get bc dof indices (local) in (V, Vg) spaces on this process that
+      // were found by other processes, e.g. a vertex dof on this process
+      // that has no connected facets on the boundary.
+      const std::vector dofs_remote = get_remote_bcs1_sub(
+          *V.dofmap()->index_map, V.dofmap()->index_map_bs(), dofs);
+
+      // Add received bc indices to dofs_local
+      dofs.insert(dofs.end(), dofs_remote.begin(), dofs_remote.end());
+
+      // Remove duplicates and sort
+      std::sort(dofs.begin(), dofs.end());
```

Comment:
It it necessary to do the sort and erase twice?

Reply:
Good question, this is done in the other implementations of `locate_dofs_topological` as well.